src/python/tritonfrontend/_api/_restricted_features.py

```diff
@@ -0,0 +1,126 @@
+# Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import json
+from copy import deepcopy
+from enum import Enum
+from typing import List, Union
+
+import tritonserver
+from pydantic import field_validator
+from pydantic.dataclasses import dataclass
+from tritonfrontend._api._error_mapping import handle_triton_error
+from tritonfrontend._c.tritonfrontend_bindings import InvalidArgumentError
+
+
+# 1-to-1 copy of RestrictedCategory Enum from https://github.com/triton-inference-server/server/blob/main/src/restricted_features.h
+class Feature(Enum):
+    HEALTH = "health"
+    METADATA = "metadata"
+    INFERENCE = "inference"
+    SHM_MEMORY = "shared-memory"
+    MODEL_CONFIG = "model-config"
+    MODEL_REPOSITORY = "model-repository"
+    STATISTICS = "statistics"
+    TRACE = "trace"
+    LOGGING = "logging"
+
+
+@dataclass
+class FeatureGroup:
+    key: str
+    value: str
+    features: List[Feature]
+
+    @field_validator("features", mode="before")
+    def validate_features(features: List[Feature]) -> List[Feature]:
+        invalid_features = [item for item in features if not isinstance(item, Feature)]
+        if invalid_features:
+            raise tritonserver.InvalidArgumentError(
```

Comment:
Currently investigating why `@field_validator` is incompatible with `@handle_triton_server` nested underneath. To unblock, directly raised `tritonserver` equivalent error.

Reply:
The function signature gets overwritten when using the `@handle_triton_error` decorator. And with the change in function signature, pydantic's `field_validator()` does not consider the underlying function it is trying to wrap to have a valid function signature. 


To fix this we use `@functools.wraps(func)` internally within `@handle_triton_error` it copies over the original function signature. In case you are curious, this is done by overwriting the function attributes of this final function. [Link to `functools.update_wrapper()` and `functools.wraps()`](https://github.com/python/cpython/blob/49f15d8667e4755876698a8daa13bab6acee5fa1/Lib/functools.py#L35C1-L77C55).